staking-pool/README.md

```diff
@@ -1,11 +1,131 @@
+The contract shouldn't have locked balance during the initialization.
+All current balance is converted to shares and will be staked (after the next action).
+This balance can never be unstaked or withdrawn from the contract.
+It's used to maintain the minimum number of shares, as well as help pay for the potentially growing contract storage.
+
+### Delegator accounts
+
+The contract maintains information per delegator in the persistent map keyed by the hash of the delegator account ID.
+
+The information contains:
+- Unstaked balance of the account.
+- Number of "Stake" shares.
+- The minimum epoch height when the unstaked balance can be withdrawn.
+
+A delegator can do the following actions:
+
+#### Deposit
+
+When a delegator account first deposit funds to the contract, the internal account is created and credited with the
+attached amount of unstaked tokens.
+
+#### Stake
+
+When an account wants to stake a given amount, the contract calculates the number of "stake" shares (`num_shares`) and the actual rounded stake amount (`amount`).
+The unstaked balance of the account is decreased by `amount`, the number of "stake" shares of the account is increased by `num_shares`.
+The contract increases the total number of staked tokens and the total number of "stake" shares. Then the contract restakes.
+
+#### Unstake
+
+When an account wants to unstake a given amount, the contract calculates the number of "stake" shares needed (`num_shares`) and
+the actual required rounded unstake amount (`amount`).
+The unstaked balance of the account is increased by `amount`, the number of "stake" shares of the account is decreased by `num_shares`.
+The minimum epoch height when the account can withdraw is set to the current epoch height increased by `4`.
+The contract decreases the total number of staked tokens and the total number of "stake" shares. Then the contract restakes.
+
+#### Withdraw
+
+When an account wants to withdraw, the contract checks the minimum epoch height of this account and checks the amount.
+Then sends the transfer and decreases the unstaked balance of the account.
+
+#### Ping
+
+Calls the internal function to distribute rewards if the blockchain epoch switched. The contract will restake in this case.
+
+### Reward distribution
+
+Before every action the contract calls method `internal_ping`.
+This method distributes rewards towards active delegators when the blockchain epoch switches.
+The rewards might be given due to staking and also because the contract earns gas fee rebates for every function call.
+
+The method first checks that the current epoch is different from the last epoch, and if it's not changed exits the method.
+
+The reward are computed the following way. The contract keeps track of the last known total account balance.
+This balance consist of the initial contract balance, and all delegator account balances (including the owner) and all accumulated rewards.
```

Comment:
>  and all accumulated rewards.

How are they added to the contract?

Reply:
Automatically.

Reply:
Can you elaborate?  If I was making a test how would this be done?  Is it a special transaction?

Reply:
account.locked just increasing from time to time. That increase is a revard (inflation). But the amount of shares is fixed to the total locked balance as if there was no inflation.